podvm: Check for pre-existing image before image creation

bpradipt opened the pull request:
For OSC, we expect only one image to be created by the operator. The IMAGE_VERSION (for Azure) or AMI_VERSION (for AWS) uses clusterID as the unique id.  This helps to check if the image already exists before recreating the image again.

<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->
**- Description of the problem which is fixed/What is the use case**

**- What I did**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

bpradipt committed:
podvm: Check for pre-existing image before image creation

For OSC, we expect only one image to be created by the operator.
The IMAGE_VERSION (for Azure, GCP) or AMI_VERSION (for AWS) uses clusterID as
the unique id.  This helps to check if the image already exists
before recreating the image again.

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

beraldoleal commented on controllers/image_generator.go:
```diff
@@ -745,15 +745,40 @@ func (r *ImageGenerator) createImageConfigMapFromFile() error {
 			igLogger.Info("Setting IMAGE_GALLERY_NAME", "image_gallery_name", image_gallery_name)
 		}
 
+		cm.Data["IMAGE_VERSION"] = imageVersion
```
Since GCP the image version will be part of the name... and the name can't have DOTs. IIUC, we should add another if here to check if the provider == GCP and set  `cm.Data["IMAGE_VERSION"] = imageVersionWithoutdots` ?

beraldoleal replied:
Since GCP the image version will be part of the name... and the name can't have DOTs. IIUC, we should add another if here to check if the provider == GCP and set  `cm.Data["IMAGE_VERSION"] = imageVersionWithoutdots` ?

bpradipt committed:
Treat azure error as non-existant image

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

beraldoleal reviewed (approved):
LGTM, thanks.

bpradipt committed:
azure: Delete leftover image from failed packer build

In certain situation there could be leftover image
due to abrupt exit of packer build.
This image is not under galleries/image-definition/image-version
and hence need to be deleted explicitly

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

beraldoleal commented:
@bpradipt thanks for handling the gcp part, saved me some time. ;)

beraldoleal commented on config/peerpods/podvm/azure-podvm-image-handler.sh:
```diff
@@ -718,6 +737,52 @@ function delete_image_using_id() {
+# 1: Image does not exist
+# 2: Image mismatch (caller should delete Image)
+# The required variables are assumed to be set
+function image_exists() {
```
I missed this method under gcp. IIUC, most of it could be generic to all providers?

bpradipt replied:
A portion of it yes, but querying is specific to individual provider. So I kept it for each provider..

beraldoleal replied:
And where is the gcp implementation? I saw you changed the gcp-podvm-image-handler.sh but I could not find any image_exists method.

bpradipt committed:
podvm: Add a preStop hook to cleanup left over packer VMs

A script to clean up left over packer resources is added and executed
as part of preStop hook in the pod vm generation job.
The packer resources uses well-defined prefix, so it makes it easier
to query the left over resources and delete them.

The logs are redirected to stdout of PID 1 so that it's accessible
via oc logs command

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

bpradipt committed:
Retry Azure gallery delete

To overcome transient error

Signed-off-by: Pradipta Banerjee <pradipta.banerjee@gmail.com>

littlejawa reviewed (approved):
/lgtm
